2021q3/wg.adoc

```diff
@@ -0,0 +1,19 @@
+
+The wireguard-freebsd repository has been geared around being an out-of-tree kmod, which is distributed in ports. But it is also organized to be eventually upstreamed. To that end, the repository maintains two files: https://git.zx2c4.com/wireguard-freebsd/tree/src/compat.h[compat.h] and https://git.zx2c4.com/wireguard-freebsd/tree/src/support.h[support.h]. compat.h contains polyfills of code that exists in FreeBSD's main branch but does not exist in various older releases, with ifdefs for each of the various releases we support. On the other hand, support.h contains code that is not *yet* in FreeBSD's main branch. The goal is to eventually move all the code from support.h into compat.h, at which point, the repository will be ready for upstreaming. As of writing, there's basically only one real function left -- `sogetsockaddr` -- and then two convenience macros that need to be sent upstream for consideration by ConcurrencyKit.
+
+A significant aspect that isn't in support.h, though, is the cryptographic primitives that the code uses. The files https://git.zx2c4.com/wireguard-freebsd/tree/src/crypto.c[crypto.c] and https://git.zx2c4.com/wireguard-freebsd/tree/src/crypto.h[crypto.h] contain boring C "reference implementations" of ChaCha20Poly1305, XChaCha20Poly1305, Blake2s, and X25519 (which is formally verified via MIT's fiat-crypto project). These four functionalities are used by the handshake path on very small inputs for WireGuard's key exchange, and will hopefully be making their way to sys/crypto/ in the FreeBSD tree as just ordinary functions. On the flip side, the datapath uses an entry point of ChaCha20Poly1305 that works on mbufs (which might be rather large) and is performance critical. To that end, jhb@ has been improving OCF for WireGuard's particulars. The next step will then be moving our current calls from `chacha20poly1305_{en,de}crypt_mbuf` in https://git.zx2c4.com/wireguard-freebsd/tree/src/wg_noise.c[wg_noise.c] to instead call out to OCF, submitting `crypto_buffer`s of type `CRYPTO_BUF_MBUF`. This will automatically benefit from Andy Polyakov's optimized ChaPoly implementations that OCF has long since imported from OpenSSL.
```

Comment:
I'd still use "features", but up to you.

Reply:
I just hesitate to change the words of the author, @zx2c4.

Reply:
Features is definitely not correct. Closer would be "primitives", but technically, ChaCha20Poly1305 is not one primitive but a particular construction of 2 and XChaCha20Poly1305 is not one primitive but a particular construction on 3! So I just went for the kind of vague "functionalities" to not really broach the topic. But I guess that came across weird.

I guess just change it to "algorithms" instead of "functionalities". After all, everything's sort of an algorithm when you're playing with computers! :-P